apps/prairielearn/src/pages/elementFiles/elementFiles.js

```diff
@@ -53,8 +57,21 @@ router.get(
     let elementFilesDir;
     if (res.locals.course) {
       // Files should be served from the course directory
-      const coursePath = chunks.getRuntimeDirectoryForCourse(res.locals.course);
-      await chunks.ensureChunksForCourseAsync(res.locals.course.id, { type: 'elements' });
+      let question_course;
+      if (req.params.variant_id) {
+        const variant = await selectVariantById(req.params.variant_id);
+        if (!variant || !idsEqual(variant.course_id, res.locals.course.id)) {
+          // the existence of the variant within the course validates that this course has sharing permissions on this question
+          throw new HttpStatusError(404, 'Not Found');
+        }
```

Comment:
There is no validation here that the user indeed has access to the variant in the first place. If a user is able to get ahold of (or guess) a variant ID that happens to be in the course they want, they would get access to the course element files. Also, I don't know how much validation would be desirable, but this would give someone access to all of the course's elements, not just the elements of the question being shared.

Reply:
> no validation here that the user indeed has access to the variant in the first place

ok I'm embarrassed that I missed this but you are right. I'm not sure what the best way to do this is. The new way seems to be doing auth within the endpoints, where the old way was using middleware. I could just add another if statement here, not sure if there is a better solution.

> this would give someone access to all of the course's elements, not just the elements of the question being shared

yes, I was aware of this one. I'm not sure if this is a problem though. A few alternatives would be: 
- have the sharing course explicitly declare somewhere which elements they are willing to share
- parse the generated variant to see which elements it is using
both of these options seem like a total pain, and not worth the trouble.

Reply:
Ideally we'd like to reuse authz functions from the middleware. I'd like to avoid hand-crafted authz checks if at all possible. We basically want to perform the usual checks for access to the question associated with the variant? And check that it's shared?

Reply:
Also, only use the variant in the URL if the question course does not match the variant course, so caching can be used and this check is not required.

Reply:
I'm not really sure what you mean by "reuse authz functions from the middleware", there aren't reusable functions, almost everything is done directly in the handlers. The closest thing to what you are saying is this: 
https://github.com/PrairieLearn/PrairieLearn/blob/04ccc99345977bfda77c20dc21e53665545d1106/apps/prairielearn/src/middlewares/selectAndAuthzInstanceQuestion.js#L11-L20

I can give it a try but it will be a little awkward because I will need to attach some things to locals for it to work, and it will throw as soon as it hits any issues.

Reply:
Discussed in a meeting, we aren't going to add these extra checks, because we aren't doing any of that for elementFiles within the same course anyway.